### PR TITLE
KEYCLOAK-2253 - Add support for ConfiguredProvider based UserFederationProviderFactory.

### PR DESCRIPTION
**Pull request by thomasdarimont**
`UserFederationProvidersResource` is now aware of `ConfiguredProvider` and allows sophisticated
configuration of configuration properties via `ProviderConfigProperty` definitions that can be provided by the `UserFederationProviderFactory` that implements the `ConfiguredProvider` interface.
See `DummyUserFederationProviderFactory. getConfigProperties()` for example.

Previously `UserFederationProvidersResource` did only support simple key-value pairs for expressing
configurable options.

Tested this by launching a standalone `KeycloakServer` and creating a new `DummyUserFederationProvider` via the admin console.
The default values, labels and help messages are correctly displayed and the values are stored correctly.

### CODE REVIEW

File: forms/common-themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js [hunk 1]
```diff
@@ -644,6 +644,16 @@ module.controller('GenericUserFederationCtrl', function($scope, $location, Notif
                 instance.config.updateProfileFirstLogin = true;
                 instance.config.allowKerberosAuthentication = true;
             }
+
+            if (providerFactory.properties) {
+
+                for (var i = 0; i < providerFactory.properties.length; i++) {
+                    var configProperty = providerFactory.properties[i];
+                    var configValue = configProperty.type == "boolean" ? (configProperty.defaultValue === true ? 'true' : 'false') : configProperty.defaultValue;
```
Comment (thomasdarimont) [inline forms/common-themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js]: Is there a better way to do this? it seems that the "on/off" toggles need the boolean as a string to function properly.

File: forms/common-themes/src/main/resources/theme/base/admin/resources/partials/federated-generic.html [hunk 2]
```diff
@@ -37,6 +37,9 @@
                     <input class="form-control" type="text" data-ng-model="instance.config[ option ]" >
                 </div>
             </div>
+
+            <kc-provider-config realm="realm" config="instance.config" properties="providerFactory.properties"></kc-provider-config>
```
Comment (thomasdarimont) [inline forms/common-themes/src/main/resources/theme/base/admin/resources/partials/federated-generic.html]: I deliberately added this kc-provder-config after the "conventional" option rendering to remain backwards compatible.

File: testsuite/integration/src/test/java/org/keycloak/testsuite/DummyUserFederationProviderFactory.java [hunk 3]
```diff
@@ -84,4 +84,29 @@ public int getFullSyncCounter() {
...
         return changedSyncCounter.get();
     }
+
+    @Override
+    public String getHelpText() {
+        return "Dummy User Federation Provider Help Text";
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
```
Comment (thomasdarimont) [inline testsuite/integration/src/test/java/org/keycloak/testsuite/DummyUserFederationProviderFactory.java]: Added default config properties for testing.

**Comment by thomasdarimont** [comment]
Would appreciate some tips on how to test this, I couldn't find a proper test class to add a new test case - at first I thought `UserFederationModelTest` was the right place but since it doesn't bootstrap a `UserFederationProviderFactory` I think another place is probably more appropriate :)

Any hints?

**Comment by patriot1burke** [comment]
I don't think we want to show a generic way to set attributes along side a custom way to do it.  It looks like that the attribute page will show two different ways to configure the same attribute

**Comment by patriot1burke** [comment]
ah, nevermind.  I see you have a new properties setting in addition to the old way

File: forms/common-themes/src/main/resources/theme/base/admin/resources/partials/federated-generic.html (hunk 2 above)
Comment (thomasdarimont) [inline forms/common-themes/src/main/resources/theme/base/admin/resources/partials/federated-generic.html]: This is how it looks:

My screenshot

**Comment by thomasdarimont** [comment]
yep :) Didn't want to break backwards compatibility :)

**Comment by patriot1burke** [comment]
What the other providers do that use ConfiguredProvider is to set a default boolean value as a string.  If you do that, then you don't need that for loop in users.js

**Comment by thomasdarimont** [comment]
Better now?

**Comment by patriot1burke** [comment]
yup, i'll merge if build passes

**Comment by thomasdarimont** [comment]
:+1:

**Pull request merged**